STM32 FDCAN Register Definitions

mattofak opened the pull request:
Register addresses and definitions taken from the STM32H7 manual.

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+/** @defgroup fdcan_defines FDCAN defines
+@ingroup STM32H7_defines
```
If you're putting this in a common directory, it can't be in the h7 file grouping please.

mattofak replied:
That's fair. Cribbing from the `can.h` file, it looks like I should put this in the `STM32F_defines` group. ST claims that this IP is not specific to the H7...

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+ * @ingroup fdcan_defines
+ * @{*/
+#define FDCAN1			FDCAN1_BASE
+#define FDCAN2			FDCAN2_BASE
```
I think these sorts of defines are normalyl in the part specific header, like h7/fdcan.h?

mattofak replied:
Happy to put them there; but it looks like `can.h` follows this same pattern of defining the device define in the file and deferring the effort to the `libopencm3/stm32/memorymap.h` header to sort things out.

Any thoughts?

mattofak replied:
Resolved this by not using can.h as a template. Started following other modules which means that this did get moved to the H7 dir and out of the common.

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+
+/**@{
+ * @defgroup fdcan_reg_base CAN register base address
+ * @ingroup fdcan_defines
```
you don't need any of these @ingroup fdcan_defines when it's the logical nesting level (like your line 1 group definition) you only need an @ingroup if you want to put it "elsewhere"

mattofak replied:
I probably should just do a doxygen and see what happens; but since I don't have a `@file` in the first doxyblock in the file, nor do I have a `@{` are we still logically nesting at this point?

karlp replied:
yes, please do run doxygen yourself and see that it looks how you expect :)  Probably easiest to open a group around the file, we do that a lot elsewhere.

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+#define FDCAN_CCCR_BRSE BIT9
+/* FD operation enable */
+#define FDCAN_CCCR_FDOE BIT8
+/* Test mode enable */
```
If you add the second * to all of these, they get doxygen too :)  (no big deal, but when you have the descriptions already)

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+};
+#define FDCAN_GFC_ANFS_SHIFT 4
+/* Accept non-matching frames extended */
+enum fdcan_gfc_anfe {
```
is there much value in these two enums being separately named? I'm not a heavy can user, would it be nicer/equally sane from an enduser point of view to just have FDCAN_GFC_ACCEPTFIFO0/1/REJECT/MASK instead of duped for anfs and anfe?

mattofak replied:
I was on the fence about this as well. I'll make them consistent.

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+/* --- FDCAN_NDAT2: FDCAN new data 2 register ------------------------------- */
+/* New data[63:32] */
+#define FDCAN_NDAT2_ND_MASK 0xFFFFFFFF
+#define FDCAN_NDAT2_ND_SHIFT 0
```
I don't really see any point to these NDAT1 and NDAT2 defines?  Is this all autogenerated?

mattofak replied:
Sort of. I wrote a helper script to pull bit definitions from a copy/paste from the reference manual. But some manual massaging had to be done.

Happy to pull these definitions out though. I wasn't sure if convention was to define everything even if it had no obvious use just to be consistent or not.

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+	FDCAN_RXESC_F0DS_48		= 0x6,
+	FDCAN_RXESC_F0DS_64		= 0x7,
+	FDCAN_RXESC_F0DS_MASK	= 0x7,
+};
```
same comment as earlier, these are all the same, it doesn't seem useful/necessary to repeat them

mattofak replied:
done

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+	FDCAN_TXESC_TBDS_48		= 0x6,
+	FDCAN_TXESC_TBDS_64		= 0x7,
+	FDCAN_TXESC_TBDS_MASK	= 0x7,
+};
```
and here

mattofak replied:
done

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+ * register  */
+/* Cancellation finished interrupt enable. */
+#define FDCAN_TXBCIE_CFIE_MASK 0xFFFFFFFF
+#define FDCAN_TXBCIE_CFIE_SHIFT 0
```
and all of these up to line 804?

mattofak replied:
Removed.

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+#define FDCAN_TXEVENT_ID_29_MASK  0x1FFFFFFF
+#define FDCAN_TXEVENT_ID_29_SHIFT 0
+#define FDCAN_TXEVENT_ID_11_MASK  0x7FF
+#define FDCAN_TXEVENT_ID_11_SHIFT 18
```
These 29 and 11 masks keep coming up too.

mattofak replied:
I thought about this for a while, and although the 11 bit and 29 bit masks are always going to be the same, I wasn't sure about how to make it clear that they belonged to the `SHIFT`.

karlp commented on include/libopencm3/stm32/fdcan.h:
```diff
@@ -0,0 +1,1420 @@
+ * address field has different meanings. */
+struct fdcan_stdfilt {
+	uint32_t flags;
+};
```
This struct doesn't seem to be doing much?

mattofak replied:
Agreed. However, the thought was that it would be clearer to use these as `struct fdcan_stdfilt[]` vs `uint32_t[]`.

karlp commented:
General idea looks fine.  Comment style is a little mixed? sometimes doxygen, sometimes not?  Do you have more of this coming, or just this header for now?

mattofak commented:
Alrighty; instead of cribbing from the existing `can.h` I started following what seemed like the convention of a bunch of other modules. So I moved the new `fdcan.h` into `common/fdcan_common.h` and added the macro magic. I also added a blank object file for fdcan so that I could compile libopencm for the H7 and see all the terrible things I did syntax wise.

mattofak commented:
Apologies for the commit/force push noise as I've apparently forgotten how to work with multiple upstreams...

karlp commented:
No problems with the pushing, this is your code, push as you see fit! :) and yeah, the existing can.h is probably not the best example of how things are normally done either.   I'd definitely recommend trying this out a bit with your own _use_ of this before we merge it as just definitions, but it's good to see it early.

karlp commented on include/libopencm3/stm32/common/fdcan_common.h:
```diff
@@ -0,0 +1,1447 @@
+/** @defgroup fdcan_defines FDCAN defines
+ * @ingroup STM32F_defines
```
this looks suspect to me, but I've not gone and checked against others.

karlp commented on lib/stm32/common/fdcan_common.c:
```diff
@@ -0,0 +1,24 @@
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2019 Brian Viele <vielster@allocor.tech>
```
This is not you, though I know you're both at allocor.tech.  Also, there's nothing _in_ this file (yet?) and you've not claimed copyright on any of the other files?

karlp replied:
This is still wrong?

mattofak commented:
@karlp I would appreciate another review. I've used this to build a simple FIFO based CAN-FD thing so it works! Controversial things I've added is a new library to ease bit timing calculations which I rewrote from the linux can-utils project; and some macros to make it easier / less error prone to write bit masking operations.

karlp commented on include/libopencm3/stm32/h7/fdcan.h:
```diff
@@ -0,0 +1,37 @@
+/** @defgroup fdcan_defines FDCAN defines
+ * @ingroup STM32F_defines
```
this doesn't look correct?

karlp commented:
Do you have any example application code you can share?  I realise it's a big peripheral, but damn, it just looks like so much noise.  I want to pull it locally and see how the docs look too, because there's lots more doxygen wrapping than I'm used to seeing :)

karlp commented:
Also, there's no "bit timing" library code here? Do you need to git push again? or anything?

karlp commented on include/libopencm3/cm3/common.h:
```diff
@@ -111,4 +111,26 @@
 #define BIT31 (1<<31)
 
+/* Generic bit mask manipulation */
+/** Set a \p field inside of \p reg_ with bits_. Assumes that the field name can
```
In general, there's no poitn in having a "non-doxygen" and then also a doxygen comment.  Just one doxygen comment.

karlp commented on include/libopencm3/cm3/common.h:
```diff
@@ -111,4 +111,26 @@
+ * @param field_ Prefix of field to extract from \p reg_
+ * @returns Value of the field inside the register.
+ */
+#define CM3_GET_MASK(reg_, field_) (((reg_) >> field_##_SHIFT) & field_##_MASK)
```
Put this stuff in a separate commit please, I'm not against it, (I mean, I am, this shit is illegible to debug, but I know it's popular) but let's have it in it's own commit so we can find it easier.  (IN this series is fine)  The USB stack uses constructs like this too, so we extract it out of there as well.

karlp commented on include/libopencm3/stm32/h7/fdcan.h:
```diff
@@ -0,0 +1,64 @@
+		.tseg1_min = 1, .tseg1_max = 32, .tseg2_min = 1,             \
+		.tseg2_max = 16, .sjw_max = 16, .brp_min = 1, .brp_max = 32, \
+		.brp_inc = 1,                                                \
+	}
```
why can't these just be const structs?  this isn't the 80s, we don't have to do everything with macro soup!

karlp replied:
(remember, data sections and garbage collection will remove unused items....)

karlp commented on lib/can/can_bit_timing.c:
```diff
@@ -0,0 +1,197 @@
+ * linux-can projects.
+ * Copyright 2004-2006 Pavel Pisa - DCE FELK CVUT cz
+ * Copyright 2005      Stanislav <pisa@cmp.felk.cvut.cz>
+ * Copyright 2019      Matt Walker <m@allocor.tech>
```
yes, this will be special.  What sort of licenses do those projects have? Are you sure you can add them to an lgpl project?

karlp commented:
I'm sorry we didn't get this worked out but some of that extra code always looked a bit special.  I believe this is now superseded by 458250dc so I'm closing it, but if there's functions here that we should still add, please feel free to reopen.
